grails-app/controllers/au/org/emii/netcdfsubset/SubsetController.groovy

```diff
@@ -21,13 +23,29 @@ class SubsetController {
         response.setContentType("application/octet-stream")
         response.setHeader("Content-disposition", "filename=${filenameToServe(subsetRequest)}")
 
-        try {
-            subsetService.subset(subsetRequest.typeName, subsetRequest.CQL_FILTER, response.outputStream)
+        if (isThreadAvailable()) {
```

Comment:
This is not necessarily thread safe and can result in undeterministic behaviour.

What you want to have is basically 2 functions `up()` and `down()` - as you do, you can keep them in the names you're using now, it's fine. Eliminate the use of `isThreadAvailable()` and play only with those two. See if `acquire()` can take any parameters so you can specify when it actually fails. You want it to fail if there are too many locks.